cluster-autoscaler/core/scale_down.go

```diff
@@ -1005,23 +1010,27 @@ func evictPod(podToEvict *apiv1.Pod, client kube_client.Interface, recorder kube
 		}
 		lastError = client.CoreV1().Pods(podToEvict.Namespace).Evict(eviction)
 		if lastError == nil || kube_errors.IsNotFound(lastError) {
-			return nil
+			return podEvictionResult{pod: podToEvict, err: nil}
 		}
 	}
 	klog.Errorf("Failed to evict pod %s, error: %v", podToEvict.Name, lastError)
 	recorder.Eventf(podToEvict, apiv1.EventTypeWarning, "ScaleDownFailed", "failed to delete pod for ScaleDown")
-	return fmt.Errorf("failed to evict pod %s/%s within allowed timeout (last error: %v)", podToEvict.Namespace, podToEvict.Name, lastError)
+	return podEvictionResult{pod: podToEvict, err: fmt.Errorf("failed to evict pod %s/%s within allowed timeout (last error: %v)", podToEvict.Namespace, podToEvict.Name, lastError)}
 }
 
 // Performs drain logic on the node. Marks the node as unschedulable and later removes all pods, giving
 // them up to MaxGracefulTerminationTime to finish.
 func drainNode(node *apiv1.Node, pods []*apiv1.Pod, client kube_client.Interface, recorder kube_record.EventRecorder,
-	maxGracefulTerminationSec int, maxPodEvictionTime time.Duration, waitBetweenRetries time.Duration) errors.AutoscalerError {
+	maxGracefulTerminationSec int, maxPodEvictionTime time.Duration, waitBetweenRetries time.Duration,
+	podEvictionHeadroom time.Duration) (err errors.AutoscalerError, resultType status.NodeDeleteResultType, failingPods []*apiv1.Pod) {
```

Comment:
Would it make sense to return map of pod names to errors here? Disconnecting the two here may be OK for aggregation purposes, but it also blocks processing them further.

Reply:
Yes, that's a good idea! Although there can still be pods that caused the node failure but don't have any specific error associated (e.g. timed out while waiting for their eviction/disappearance), so I'd leave both fields actually.